### PR TITLE
Moved ceased bwt undo ops to ad-hoc structure

### PR DESCRIPTION
**Pull request by abi87**
@alsala , please review this but let's cross-check together when we can merge it

### CODE REVIEW

File: src/coins.cpp [hunk 1]
```diff
@@ -1330,22 +1330,24 @@ bool CCoinsViewCache::HandleCeasingScs(int height, CBlockUndo& blockUndo)
                 continue;
 
             if (!foundFirstBwt) {
-                blockUndo.vtxundo.push_back(CTxUndo());
-                blockUndo.vtxundo.back().refTx = scInfo.lastCertificateHash;
-                blockUndo.vtxundo.back().firstBwtPos = pos;
+                blockUndo.vVoidedCertUndo.push_back(CVoidedCertUndo());
+                blockUndo.vVoidedCertUndo.back().voidedCertHash = scInfo.lastCertificateHash;
+                blockUndo.vVoidedCertUndo.back().firstBwtPos = pos;
                 LogPrint("cert", "%s():%d - set refTx[%s], pos[%d]\n",
```
Comment (alsala) [inline src/coins.cpp]: In the log refTx is a refuse

File: src/main.cpp [hunk 2]
```diff
@@ -2313,13 +2313,13 @@ bool DisconnectBlock(CBlock& block, CValidationState& state, CBlockIndex* pindex
     LogPrint("sc", "%s():%d - ===============> CBlockUndo red from DB:\n%s\n",
         __func__, __LINE__, blockUndo.ToString());
     // no coinbase in blockundo
-    if (blockUndo.vtxundo.size() < (block.vtx.size() - 1 + block.vcert.size()))
+    if (blockUndo.vtxundo.size() != (block.vtx.size() - 1 + block.vcert.size()))
         return error("DisconnectBlock(): block and undo data inconsistent");
 
-    for(int idx = block.vtx.size() - 1 + block.vcert.size(); idx < blockUndo.vtxundo.size(); idx++)
+    for(int idx = 0; idx < blockUndo.vVoidedCertUndo.size(); idx++)
```
Comment (alsala) [inline src/main.cpp]: Are they to be processed in reverse order as in the loops below?

File: src/undo.h [hunk 3]
```diff
@@ -193,9 +175,10 @@ class CBlockUndo
     bool includesSidechainAttributes;
 
 public:
-    std::vector<CTxUndo> vtxundo; // for all but the coinbase
+    std::vector<CTxUndo>                vtxundo;         // for all txs and certs but the coinbase
     uint256 old_tree_root;
-    std::map<uint256, ScUndoData> msc_iaundo; // key=scid, value=amount matured at block height
+    std::vector<CVoidedCertUndo> vVoidedCertUndo; // for all but the coinbase
```
Comment (alsala) [inline src/undo.h]: just to be formally consistent with the serialization order they could be swapped

**Review by alsala — APPROVED** [review]

File: src/coins.cpp (hunk 1 above)
Comment (abi87) [reply src/coins.cpp]: Done

File: src/undo.h (hunk 3 above)
Comment (abi87) [reply src/undo.h]: Done

File: src/main.cpp (hunk 2 above)
Comment (abi87) [reply src/main.cpp]: Done

**Pull request merged**